File: 3939-ProcessStringWithSpecialOperationsIi/3939-ProcessStringWithSpecialOperationsIi.py
# Last updated: 7/9/2026, 10:06:26 AM
class Solution:
    def processStr(self, s: str, k: int) -> str:
        n = len(s)

        length = [0] * (n + 1)

        for i, ch in enumerate(s):
            if ch.isalpha():
                length[i + 1] = length[i] + 1
            elif ch == '*':
                length[i + 1] = max(0, length[i] - 1)
            elif ch == '#':
                length[i + 1] = length[i] * 2
            else:  # '%'
                length[i + 1] = length[i]

        if k >= length[n]:
            return '.'

        for i in range(n - 1, -1, -1):
            ch = s[i]
            prev_len = length[i]

            if ch.isalpha():
                if k == prev_len:
                    return ch

            elif ch == '#':
                if k >= prev_len:
                    k -= prev_len

            elif ch == '%':
                k = prev_len - 1 - k

            # '*' does not change k when tracing backwards

        return '.'

File: 3939-ProcessStringWithSpecialOperationsIi/3939-ProcessStringWithSpecialOperationsIi.java
// Last updated: 7/9/2026, 10:06:43 AM
class Solution {
    public char processStr(String s, long k) {
        int n = s.length();
        long[] len = new long[n + 1];

        for (int i = 0; i < n; i++) {
            char c = s.charAt(i);

            if (c >= 'a' && c <= 'z') {
                len[i + 1] = len[i] + 1;
            } else if (c == '*') {
                len[i + 1] = Math.max(0, len[i] - 1);
            } else if (c == '#') {
                len[i + 1] = len[i] * 2;
            } else { // %
                len[i + 1] = len[i];
            }
        }

        if (k >= len[n]) {
            return '.';
        }

        for (int i = n - 1; i >= 0; i--) {
            char c = s.charAt(i);
            long prevLen = len[i];

            if (c >= 'a' && c <= 'z') {
                if (k == prevLen) {
                    return c;
                }
            } else if (c == '#') {
                if (k >= prevLen) {
                    k -= prevLen;
                }
            } else if (c == '%') {
                k = prevLen - 1 - k;
            }
            // '*' needs no change to k
        }

        return '.';
    }
}